File: 1235-maximum-profit-in-job-scheduling/1235-maximum-profit-in-job-scheduling.cpp
bool comp(const vector<int> &a, const vector<int> &b) {
    return (a[0] <= b[0]); // compare as per start time
}

class Solution {
public:
    int n;
    vector<vector<int>> arr;
    vector<int> dp;

    int find(int i) {
        if (i >= n) {
            return 0;
        }
        if (dp[i] != -1) {
            return dp[i];
        }
        vector<int> t = { arr[i][1], -1, -1 }; // start, end and profit respectively
        int id = lower_bound(arr.begin(), arr.end(), t) - arr.begin();

        //we either take the current index or return
        int ans = arr[i][2] + find(id);
        ans = max(ans, find(i + 1));
        return dp[i] = ans;
    }

    int jobScheduling(vector<int>& startTime, vector<int>& endTime, vector<int>& profit) {
        n = startTime.size();
        for (int i = 0; i < n; i++) {
            arr.push_back({ startTime[i], endTime[i], profit[i] });
        }
        dp.resize(n, -1);
        sort(arr.begin(), arr.end(), comp);
        return find(0);
    }
};

File: 1235-maximum-profit-in-job-scheduling/1235-maximum-profit-in-job-scheduling.java
class Solution {
    public int n;
    public int[][] arr;
    public int[] dp;
    
    public int jobScheduling(int[] startTime, int[] endTime, int[] profit) {
        n = startTime.length;
        arr = new int[n][3];
        for (int i = 0; i < n; i++) {
            arr[i][0] = startTime[i];
            arr[i][1] = endTime[i];
            arr[i][2] = profit[i];
        }
        dp = new int[n];
        Arrays.fill(dp, -1);
        Arrays.sort(arr, (a, b) -> {
            return a[0] - b[0];
        });
        return solve(0);
    }
    
    public int solve(int i) {
        if (i >= n) {
            return 0;
        }
        if (dp[i] != -1) {
            return dp[i];
        }
        int id = lowerBound(arr[i][1]);
        return dp[i] = Math.max(
            arr[i][2] + solve(id), // take
            solve(i + 1) // not take
        );
    }
    
    public int lowerBound(int temp) {
        int low = 0, high = n - 1, res = 0;
        while (low <= high) {
            int mid = (high + low) / 2;
            if (arr[mid][0] >= temp) {
                high = mid - 1;
                res = mid;
            } else {
                low = mid + 1;
                res = mid + 1;
            }
        }
        return res;
    }
}